### ISSUE TITLE
Test performances issues when trying to verify app navigation workflow (testing urls)

### ISSUE THREAD

**Issue by Vadorequest** [cypress-io/cypress#8510]
### Current behavior:

I'm running 2 specs, `footer` and `nav`. Footer works, while nav fails on test N°3 `should have a link in the navbar that redirects to the built-in feature "Static I18n"`.

The reason why it fails is directly related to the timeout defined within the test:
```js
cy.url({ timeout: 1000 }).should('eq', `${baseUrl}/en/examples/built-in-features/static-i18n`);
```

> I'm currently debbuging https://github.com/UnlyEd/next-right-now/pull/159 (full github action log at https://github.com/UnlyEd/next-right-now/runs/1077454627 with the whole tests suite)

Those tests used to pass (with timeout 10000 instead of 1000), but since recently, they don't pass anymore. 

Locally, if I use a timeout of 15000 for both tests (N°3 & N°4), it seems to pass. BUT:
- It fails on Github Actions regardless, even with 15s timeout
- Even 10s timeout is SO MUCH WAITING, I don't understand why it would take 15 seconds to test something so "simple", it's basically a show/hide of a nav menu that displays links and a click on one of those links then wait for the redirection and check for the new url
- How can those tests be reliable if they depend on a variable timeout that we can't foreseen? It can break anytime
- The error when the timeout is reached is really not developer-friendly, I didn't understand that was the issue until trying to reproduce the tests locally. The error thrown by Cypress seems to be https://on.cypress.io/uncaught-exception-from-application but even though I added `uncaught:exception` listener locally (didn't push that code yet) it didn't seem to be related to the local issue, as increasing the timeout fixed the issue locally. (I'll try pushing it, see if it helps with debugging on Github Actions)

I'm concerned **I'm either doing it wrong** and don't know how to reliably test that simple behavior, or something is wrong with Cypress itself because it shouldn't take this long to run such basic tests.

#### Cypress artifacts:
- Videos: https://github.com/UnlyEd/next-right-now/suites/1150526110/artifacts/16580593
- Screenshots: https://github.com/UnlyEd/next-right-now/suites/1150526110/artifacts/16580592

#### Logs when running the specs locally:

```
$ CYPRESS_STAGE=${CYPRESS_STAGE:-development}; cypress run --config-file cypress/config-$CYPRESS_STAGE.json --spec 'cypress/integration/app/common/*.js'

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    4.8.0                                                                              │
  │ Browser:    Electron 80 (headless)                                                             │
  │ Specs:      2 found (app/common/footer.js, app/common/nav.js)                                  │
  │ Searched:   cypress/integration/app/common/*.js                                                │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  app/common/footer.js                                                            (1 of 2)


  Common > Footer section
    ✓ should have the Unly logo in the footer (7730ms)
    ✓ should have the customer logo in the footer (65ms)
    ✓ should have a button to change the language which changes the language upon click (2835ms)


  3 passing (11s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        3                                                                                │
  │ Passing:      3                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     10 seconds                                                                       │
  │ Spec Ran:     app/common/footer.js                                                             │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /Users/vadorequest/dev/next-right-now/cypress/videos/develo     (1 second)
                          pment/app/common/footer.js.mp4                                            


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  app/common/nav.js                                                               (2 of 2)


  Common > Nav section
    ✓ should have 5 links in the navigation bar (5221ms)
    ✓ should have a link in the navbar that redirects to the home page (5470ms)
    1) should have a link in the navbar that redirects to the built-in feature "Static I18n"
    ✓ should have a link in the navbar that redirects to the native feature "SSR" (14801ms)


  3 passing (32s)
  1 failing

  1) Common > Nav section
       should have a link in the navbar that redirects to the built-in feature "Static I18n":

      Timed out retrying
      + expected - actual

      -'http://localhost:8888/en'
      +'http://localhost:8888/en/examples/built-in-features/static-i18n'
      
      at Context.eval (http://localhost:8888/__cypress/tests?p=cypress/integration/app/common/nav.js:26:35)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        4                                                                                │
  │ Passing:      3                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     32 seconds                                                                       │
  │ Spec Ran:     app/common/nav.js                                                                │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /Users/vadorequest/dev/next-right-now/cypress/screenshots/development/app/common     (1280x720)
     /nav.js/Common  Nav section -- should have a link in the navbar that redirects t               
     o the built-in feature Static I18n (failed).png                                                


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /Users/vadorequest/dev/next-right-now/cypress/videos/develo    (3 seconds)
                          pment/app/common/nav.js.mp4                                               


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  app/common/footer.js                     00:10        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  app/common/nav.js                        00:32        4        3        1        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 2 failed (50%)                      00:43        7        6        1        -        -  

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.


```

### Test code to reproduce

Here is the current code in `cypress/integration/app/common/nav.js`.

> Note the `{ timeout: 1000 }` on the 3rd test, and `{ timeout: 15000 }` on the 4th.

```js
const baseUrl = Cypress.config().baseUrl;

describe('Common > Nav section', () => {
  /*
  * Visits the home page before any test
  */
  beforeEach(() => {
    cy.visit('/en');
  });

  it('should have 5 links in the navigation bar', () => {
    cy.get('#nav .navbar-nav > .nav-item').should('have.length', 5);
  });

  it('should have a link in the navbar that redirects to the home page', () => {
    cy.get('#nav-link-home')
      .should('have.text', 'Home')
      .click();
    cy.url({ timeout: 10000 }).should('eq', `${baseUrl}/en`);
  });

  it('should have a link in the navbar that redirects to the built-in feature "Static I18n"', () => {
    cy.get('#nav-link-examples')
      .should('have.text', 'Examples')
      .click();
    cy.get('#nav-link-examples-static-i-18-n')
      .should('have.text', 'Static i18n')
      .click();
    cy.url({ timeout: 1000 }).should('eq', `${baseUrl}/en/examples/built-in-features/static-i18n`);
    cy.get('h1').should('have.length', 1).should('have.text', 'Static i18n examples, using i18next and Locize vendor');
  });

  it('should have a link in the navbar that redirects to the native feature "SSR"', () => {
    cy.get('#nav-link-examples')
      .should('have.text', 'Examples')
      .click();
    cy.get('#nav-link-examples-ssr-get-server-side-props')
      .should('have.text', 'SSR (getServerSideProps)')
      .click();
    cy.url({ timeout: 15000 }).should('eq', `${baseUrl}/en/examples/native-features/example-with-ssr`);
    cy.get('h1').should('have.length', 1).should('have.text', 'Example, using SSR');
  });
});

```

#### Reproducing locally

This is an OSS project that you can clone yourself by doing the following:
- `git clone https://github.com/UnlyEd/next-right-now.git nrn-quick-start`
- `cd nrn-quick-start && git checkout bugfix-157-e2e-failing`
- `yarn`
- `yarn start`
- `yarn e2e:run:spec:common` (on another terminal)

### Versions

4.8.0, MacOS, Electron 80

**Comment by Vadorequest** [comment]
Attempt to use a `45000` ms timeout for each test didn't work on GHA, but it did the trick locally.

See https://github.com/UnlyEd/next-right-now/runs/1077495797

![image](https://user-images.githubusercontent.com/3807458/92322529-19329d80-f032-11ea-8e2b-1cfc8230093d.png)

**Comment by Vadorequest** [comment]
Adding `uncaught:exception` made the tests pass on GHA. (commit: https://github.com/UnlyEd/next-right-now/pull/159/commits/fce8df6fd8686d8735bf43e0c2283b52f9215f58)

https://github.com/UnlyEd/next-right-now/runs/1077505814

![image](https://user-images.githubusercontent.com/3807458/92322767-87c42b00-f033-11ea-97b8-33d67a1d5dd4.png)

So, the initial issue is fixed (tests are passing now), but the underlying issue isn't (I don't know what failed, I know I fixed it using `uncaught:exception` but I don't know what application error I silenced in the process, and despite adding some log using `console.error('Application error caught:', err, runnable);`, I can't find any in the Cypress logs on GHA.

> So, I fixed something that I know nothing of, by doing something that doesn't tell me what was wrong in the first place. **That's not a great developer experience, and that's what this issue is really about. Improving DX for this kind of issue 😉**

```js
Cypress.on('uncaught:exception', (err, runnable) => {
  // returning false here prevents Cypress from
  // failing the test
  console.error('Application error caught:', err, runnable); // Don't see any log of this
  return false;
});
```

**Comment by freddydumont** [comment]
I'm getting a similar problem, also testing navigation in Next.js and also only in CI.

I just tested downgrading from Next.js 9.5.3 to 9.5.2 and tests are passing now.

**Comment by Vadorequest** [comment]
I experienced this on Next 9.4.4. 
Since I fixed it, I upgraded to 9.5.4-canary.4, and they're still passing.
